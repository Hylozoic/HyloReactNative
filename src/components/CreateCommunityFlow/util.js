export const slugValidatorRegex = /^[0-9a-z-]{2,40}$/
export const invalidSlugMessage = 'URLs must have between 2 and 40 characters, and can only have lower case letters, numbers, and dashes.'
