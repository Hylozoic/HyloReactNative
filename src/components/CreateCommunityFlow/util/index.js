export const slugValidatorRegex = /^[0-9a-z-]{2,40}$/
export const invalidSlugMessage = 'URLs must have between 2 and 40 characters, and can only have lower case letters, numbers, and dashes.'

export function formatDomainWithUrl (communityDomain) {
  if (!communityDomain) return
  let formattedDomain = communityDomain.replace(/hylo\.com\/c\/?/, '')
  if (formattedDomain !== '') {
    formattedDomain = 'hylo.com/c/' + formattedDomain
  }
  return formattedDomain
}

export function removeUrlFromDomain (communityDomain) {
  if (!communityDomain) return
  return communityDomain.replace('hylo.com/c/', '')
}
