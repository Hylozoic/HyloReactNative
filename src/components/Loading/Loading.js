import React from 'react'
import { View, ActivityIndicator } from 'react-native'

export default function Loading ({ style }) {
  return (
    <View style={[style, styles.container]}>
      <ActivityIndicator />
    </View>
  )
}

export function LoadingScreen () {
  return <Loading style={{ flex: 1 }} />
}
LoadingScreen.navigationOptions = { headerShown: false }

const styles = {
  container: {
    justifyContent: 'space-around'
  },
  text: {
    textAlign: 'center'
  }
}
