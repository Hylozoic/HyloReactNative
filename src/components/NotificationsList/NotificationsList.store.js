import { createSelector } from 'redux-orm'
import orm from '../../store/models'

export const FETCH_NOTIFICATIONS = 'NotificationsList/FETCH_NOTIFICATIONS'

export const getNotifications = createSelector(
  orm,
  state => state.orm,
  session => {
    return session.Notification.all().toRefArray()
  }
)

export function fetchNotifications (first = 20) {
  return {
    type: FETCH_NOTIFICATIONS,
    graphql: {
      query: `query ($first: Int) {
        notifications (first: $first, order: "desc", resetCount: true) {
          total
          hasMore
          items {
            id
            createdAt
            activity {
              id
              actor {
                id
                name
                avatarUrl
              }
              comment {
                id
                text
              }
              post {
                id
                title
              }
              community {
                id
                name
                slug
              }
              meta {
                reasons
              }
              action
              unread
            }
          }
        }
      }`,
      variables: {
        first
      }
    },
    meta: {
      extractModel: 'Notification',
      resetCount: true
    }
  }
}
