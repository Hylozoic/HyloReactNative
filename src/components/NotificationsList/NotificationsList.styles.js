import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  notificationsList: {
    backgroundColor: 'white'
  }
})
