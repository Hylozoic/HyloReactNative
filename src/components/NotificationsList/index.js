import { withNavigationFocus } from 'react-navigation-is-focused-hoc'

import component from './NotificationsList'
import connector from './NotificationsList.connector'

export default connector(withNavigationFocus(component))
