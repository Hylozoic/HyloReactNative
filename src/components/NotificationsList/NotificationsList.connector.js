import { connect } from 'react-redux'
import React from 'react'
import { isEmpty } from 'lodash'
import {
  FETCH_NOTIFICATIONS,
  fetchNotifications,
  getNotifications,
  getHasMoreNotifications,
  markActivityRead,
  markAllActivitiesRead,
  updateNewNotificationCount
} from './NotificationsList.store'
import getMemberships from '../../store/selectors/getMemberships'
import { HeaderButton } from 'navigation/header'
import { mapWhenFocused } from 'util/connector'

const NOTIFICATIONS_PAGE_SIZE = 20

export function mapStateToProps (state, props) {
  return {
    hasMore: getHasMoreNotifications(state),
    pending: state.pending[FETCH_NOTIFICATIONS],
    notifications: getNotifications(state, props),
    currentUserHasMemberships: !isEmpty(getMemberships(state))
  }
}

export function mapDispatchToProps (dispatch, { navigation }) {
  const right = {
    onPress: () => dispatch(markAllActivitiesRead()),
    text: 'Mark as read'
  }
  return {
    fetchNotifications: () => dispatch(fetchNotifications(NOTIFICATIONS_PAGE_SIZE)),
    fetchMore: offset => dispatch(fetchNotifications(NOTIFICATIONS_PAGE_SIZE, offset)),
    markActivityRead: id => dispatch(markActivityRead(id)),
    setRightButton: () => navigation.setParams({ headerRight: <HeaderButton {...right} /> }),
    updateNewNotificationCount: () => dispatch(updateNewNotificationCount()),
    goToCreateCommunityName: () => {
      navigation.navigate('CreateCommunityName')
    }
  }
}

export default connect(
  mapWhenFocused(mapStateToProps),
  mapWhenFocused(mapDispatchToProps)
)
