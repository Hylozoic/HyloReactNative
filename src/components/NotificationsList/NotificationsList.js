import React, { Component } from 'react'
import { FlatList, TouchableOpacity, View, Text } from 'react-native'

import header from 'util/header'
import Loading from '../Loading'
import NotificationCard from '../NotificationCard'
import styles from './NotificationsList.styles'

export default class NotificationsList extends Component {
  static navigationOptions = ({ navigation }) =>
    header(navigation, { left: 'close', title: 'Notifications' })

  componentDidMount () {
    const { fetchNotifications, setRightButton, updateNewNotificationCount } = this.props
    fetchNotifications()
    updateNewNotificationCount()
    setRightButton()
  }

  fetchMore = offset => this.props.fetchMore(offset)

  keyExtractor = item => item.id

  render () {
    const { hasMore, markActivityRead, notifications, pending } = this.props
    if (pending && notifications.length === 0) return <Loading />
    if (!pending && notifications.length === 0) {
      return <Text style={styles.center}>Nothing new for you!</Text>
    }

    return <View style={styles.notificationsList}>
      <FlatList
        data={notifications}
        keyExtractor={this.keyExtractor}
        onEndReached={hasMore ? () => this.fetchMore(notifications.length) : null}
        renderItem={({ item }) =>
          <NotificationRow
            markActivityRead={markActivityRead}
            notification={item} />} />
    </View>
  }
}

export function NotificationRow ({ markActivityRead, notification }) {
  return <View>
    <TouchableOpacity onPress={() => {
      if (notification.unread) markActivityRead(notification.activityId)
      notification.onPress()
    }}>
      <NotificationCard notification={notification} />
    </TouchableOpacity>
  </View>
}
