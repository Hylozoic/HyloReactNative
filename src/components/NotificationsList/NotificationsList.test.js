import { TouchableOpacity } from 'react-native'
import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import TestRenderer from 'react-test-renderer'
import { simulate } from 'util/testing'

import Loading from '../Loading'
import NotificationsList, { NotificationRow } from './NotificationsList'

describe('NotificationsList', () => {
  let props = null
  let shallowRenderer = null

  beforeEach(() => {
    shallowRenderer = new ShallowRenderer()
    props = {
      fetchMore: () => {},
      fetchNotifications: () => {},
      hasMore: true,
      markActivityRead: jest.fn(),
      pending: false,
      notifications: [
        {
          activityId: "1",
          actor: { avatarUrl: "https://wombat.com", name: "Wombat McAardvark" },
          avatarSeparator: false,
          createdAt: "2 min ago",
          id: "1",
          onPress: jest.fn(),
          unread: false
        }
      ],
      setRightButton: () => {}
    }
  })

  it('matches the last snapshot', () => {
    shallowRenderer.render(<NotificationsList { ...props } />)
    const actual = shallowRenderer.getRenderOutput()

    expect(actual).toMatchSnapshot()
  })

  it('returns Loading with no notifications when pending is true', () => {
    props.pending = true
    props.notifications = []
    shallowRenderer.render(<NotificationsList { ...props } />)
    const actual = shallowRenderer.getRenderOutput()

    expect(actual).toEqual(<Loading />)
  })

  it('marks the notification read on touch', () => {
    const notification = props.notifications[0]
    notification.unread = true
    const root = TestRenderer.create(<NotificationsList { ...props } />).root
    simulate(root.findAllByType(TouchableOpacity)[0], 'press')
    expect(props.markActivityRead).toHaveBeenCalledWith(notification.activityId)
  })

  it('calls notification.onPress on touch', () => {
    const notification = props.notifications[0]
    const root = TestRenderer.create(<NotificationsList { ...props } />).root
    simulate(root.findAllByType(TouchableOpacity)[0], 'press')
    expect(notification.onPress).toHaveBeenCalled()
  })

  it('matches the last snapshot for NotificationRow', () => {
    shallowRenderer.render(<NotificationRow
      markActivityRead={props.markActivityRead}
      notification={props.notifications[0]} />)
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot()
  })
})
