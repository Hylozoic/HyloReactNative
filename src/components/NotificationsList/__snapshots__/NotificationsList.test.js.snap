// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsList matches the last snapshot 1`] = `
<View>
  <RCTScrollView
    data={
      Array [
        Object {
          "activityId": "1",
          "actor": Object {
            "avatarUrl": "https://wombat.com",
            "name": "Wombat McAardvark",
          },
          "avatarSeparator": false,
          "createdAt": "2 min ago",
          "id": "1",
          "onPress": [Function],
          "unread": false,
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <View>
              <View>
                <View>
                  <Image
                    source={
                      Object {
                        "uri": "https://wombat.com",
                      }
                    }
                  />
                </View>
              </View>
              <View>
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  numberOfLines={2}
                >
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                  />
                </Text>
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  numberOfLines={2}
                >
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                  >
                    Wombat 
                  </Text>
                </Text>
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                >
                  2 min ago
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`NotificationsList matches the last snapshot for NotificationRow 1`] = `
<View>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <NotificationCard
      notification={
        Object {
          "activityId": "1",
          "actor": Object {
            "avatarUrl": "https://wombat.com",
            "name": "Wombat McAardvark",
          },
          "avatarSeparator": false,
          "createdAt": "2 min ago",
          "id": "1",
          "onPress": [Function],
          "unread": false,
        }
      }
    />
  </TouchableOpacity>
</View>
`;
